processing/src/main/java/org/apache/druid/segment/join/JoinableFactory.java

```diff
@@ -43,8 +43,18 @@
    *
    * @param dataSource the datasource to join on
    * @param condition  the condition to join on
-   *
    * @return a Joinable if this datasource + condition combo is joinable; empty if not
    */
   Optional<Joinable> build(DataSource dataSource, JoinConditionAnalysis condition);
+
+  /**
+   * Compute the cache key for a data source participating in join operation. This is done separately from {{@link #build(DataSource, JoinConditionAnalysis)}}
+   * which can be an expensive operation and can potentially be avoided if cached results can be used.
+   *
+   * @param dataSource the datasource to join on
+   */
+  default Optional<byte[]> computeJoinCacheKey(DataSource dataSource)
```

Comment:
hmm. I do believe this condition should be outside. In some cases, when one of the conditions cannot be hash-joined, we can skip building the other expensive Joinable objects themselves when query involve multiple joins. That being said, I understand that we do not want to change the interfaces right now. Though when new join algorithms come into being, I believe these interfaces will change as you have to somehow pass what join algorithm you intend to use.

Reply:
Good point. I do agree that it could be better to check the condition outside so that we can stop creating potentially expensive joinables early. However, it doesn't mean that this interface should not match to `build()`. So, I would suggest either adding a description in the Javadoc why those interfaces don't match or modifying this interface to match since it's confusing when you see this code without knowing why they are different even though they seem to be supposed to match.

Reply:
+1 for having a consistent interface with `build` since they need to make the same decisions. I'm not sure this PR is the correct venue to refactor `build` so I would be in favor of adding `JoinConditionAnalysis` argument

Reply:
Thank you for your suggestions. I agree that functions should have the parity. I changed the code accordingly.